loop3d community profile HTML & URIs

nicholascar opened the pull request:
This PR implements:
- HTML presentation of the Loop3D profile for easier use
    - HTML created automatically from the RDF using [pyLODE](https://github.com/rdflib/pyLODE/)
- all items placed into a folder within `community/` for easier management

**URIs**
Please could the following URIs be enabled, as per the profile.ttl file:

- (Profile) <http://qudt.org/community/loop3d> &rarr; 
    - HTML &rarr; <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.html>
    - RDF (turtle) &rarr; <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.ttl>
- (Vocab) <http://qudt.org/community/loop3d/voc>
    - HTML &rarr; <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/voc.html>
    - RDF (turtle) &rarr; <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/voc.ttl>

The redirects could be of the form, if using Apache:

```
RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule ^/community/loop3d$           https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.ttl [R=302,L]
RewriteRule ^/community/loop3d.ttl$       https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.ttl [R=302,L]
RewriteRule ^/community/loop3d$           https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.html [R=302,L]
```

... and the above repeated also for the vocab.

(Happy to help test all redirects)

steveraysteveray commented:
Nicholas,
I have implemented a slight variation on your redirects that mostly seem to work. Please check. (The voc link seems to want to go to the .ttl, while the profile link goes to .html. Not sure why)

Also, I was planning on adding the following on our qudt.org homepage, with the two URLs you asked to be redirected. Is this useful? Would you prefer supplying some better landing page?

![image](https://user-images.githubusercontent.com/1130189/103957332-5a7dab80-50ff-11eb-843e-59cdd97f5b8b.png)

nicholascar commented:
Hi Steve,

Thanks very much for implementing the redirects and for offering to
highlight the Community Resources! I will get you more Community Resources
over time - communities such as the Geology people - to make that section
look like it's thriving.

*Redirection targets*
An apology from me: I realise I gave you poor targets for the redirects! In
the PR I indicated this:


   - (Profile) http://qudt.org/community/loop3d →
      - HTML →
      https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.html
      - RDF (turtle) →
      https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.ttl
   - (Vocab) http://qudt.org/community/loop3d/voc
      - HTML →
      https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/voc.html
      - RDF (turtle) →
      https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/voc.ttl

But really those are bad targets for redirects as they are the resources as
presented for editing in GitHub, not the rendered content, the HTML & RDF.
I should have given you this:


   - (Profile) http://qudt.org/community/loop3d →
      - HTML →
      https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.html
      <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.html>
      - RDF (turtle) →
      https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl
      <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/profile.ttl>
   - (Vocab) http://qudt.org/community/loop3d/voc
      - HTML →
      https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.html
      <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/voc.html>
      - RDF (turtle) →
      https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl
      <https://github.com/qudt/qudt-public-repo/blob/master/community/loop3d/voc.ttl>

These links actually render the content properly, with proper Content-Type
headers for the RDF files too.

So, the updated Apache config should be:

RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule ^/community/loop3d$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl
[R=302,L]
RewriteRule ^/community/loop3d.ttl$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl
[R=302,L]
RewriteRule ^/community/loop3d$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.html
[R=302,L]

RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule ^/community/loop3d/voc$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl
[R=302,L]
RewriteRule ^/community/loop3d/voc.ttl$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl
[R=302,L]
RewriteRule ^/community/loop3d/voc$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.html
[R=302,L]

I've implemented these using the testing PID domain test.linked.data.gov.au
so we can work out any issues, see:

http://test.linked.data.gov.au/community/loop3d
curl -L -H 'Accept: text/turtle'
http://test.linked.data.gov.au/community/loop3d
http://test.linked.data.gov.au/community/loop3d.ttl
http://test.linked.data.gov.au/community/loop3d/voc
curl -L -H 'Accept: text/turtle'
http://test.linked.data.gov.au/community/loop3d/voc
http://test.linked.data.gov.au/community/loop3d/voc.ttl


*Community Resources*
Again, thanks for offering to highlight these on the QUDT home page. Can I
ask for a slight change in the presentation format though? Rather than just:

---
Community Resources

   - Loop3D Profile
   - Loop3D Vocab

---

Can we instead have:

---
Community Resources

   - Loop3D Profile
      - Loop3D Vocabulary of Units

---

It may seem small but this indentation will indicate that the Vocab (of
Units, as opposed to others in the future such as a QK vocab we may make)
is part of the Profile, and I think that's important. Of course if the
links actually use the PIDs with the updated redirection targets above, the
link to the Profile will allow a user to see that the vocab is part of the
profile.

Thanks Steve, appreciate your work here!

Nick



On Fri, Jan 8, 2021 at 9:45 AM steveraysteveray <notifications@github.com>
wrote:

> Nicholas,
> I have implemented a slight variation on your redirects that mostly seem
> to work. Please check. (The voc link seems to want to go to the .ttl, while
> the profile link goes to .html. Not sure why)
>
> Also, I was planning on adding the following on our qudt.org homepage,
> with the two URLs you asked to be redirected. Is this useful? Would you
> prefer supplying some better landing page?
>
> [image: image]
> <https://user-images.githubusercontent.com/1130189/103957332-5a7dab80-50ff-11eb-843e-59cdd97f5b8b.png>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qudt/qudt-public-repo/pull/330#issuecomment-756454888>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABX3SECJJFJPLRNKSHSGZ4TSYZBTPANCNFSM4VKMFSPQ>
> .
>


-- 

______________________________________________________________________________________
kind regards
*Dr Nicholas Car*
Data Systems Architect at SURROUND Australia Pty Ltd
Address  Level 9, Nishi Building,
                  2 Phillip Law Street
                  New Acton Canberra 2601
Phone     +61 477 560 177 <++61+477+560+177>
Email       nicholas.car@surroundaustralia.comWebsite
https://www.surroundaustralia.com

*Enhancing Intelligence Within Organisations*
*delivering evidence that connects decisions to outcomes*


[image: Australian-National-University-Logo-1 – ANU Centre for Water and
Landscape Dynamics]

*Dr Nicholas Car*
Adj. Senior Lecturer

Research School of Computer Science

The Australian National University
Canberra ACT Australia



 https://orcid.org/0000-0002-8742-7730

https://cs.anu.edu.au/people/nicholas-car

steveraysteveray commented:
Nicholas,
As you may have noticed with Issue #334, simply adding your lines to my .htaccess file at the qudt root level broke resolution of things like qudt.org/2.1/vocab/unit. You also know that I'm a neophyte at Apache hacking. Here's what I did to avoid messing up what we already have.

I made an explicit "community" folder and put the following lines in the local .htaccess file there instead:

```
RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule ^/loop3d$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl
[R=302,L]
RewriteRule ^/loop3d.ttl$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl
[R=302,L]
RewriteRule ^/loop3d$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.html
[R=302,L]

RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule ^/loop3d/voc$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl
[R=302,L]
RewriteRule ^/loop3d/voc.ttl$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl
[R=302,L]
RewriteRule ^/loop3d/voc$
https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.html
[R=302,L]
```

...which results in an Internal Server Error.
Also tried it without the leading / on the RewriteRule lines. Still an error.
So, I turn to you for help.

Also, in reading the documentation on githack, we should eventually be using the cdn URL out of respect for Pavel Puchkin, right?

nicholascar commented:
> adding your lines to my .htaccess file at the qudt root level broke resolution of things like qudt.org/2.1/vocab/unit

OK, well because I can't see the entire qudt.org redirection rules, I can't know if it will break things! I suppose I could guess based on my knowledge of what qudt.org has implemented but best is just to test it.

> I made an explicit "community" folder...

I think this approach is fine and it's correct to remove the leading "/" from each of the redirect patterns because you are already "inside" `community/` when the .htaccess file is triggered.

I can't say why you've encountered a server error unfortunately. There's noting unusual with the patterns I've supplied. They are just triggering on patterns and a couple of HTTP variables (the query strings and `Accept` headers) which is all normal stuff.

I always use Apache vhost files for redirects (see Aust Govt Linked Data WG's mail PID redirect file at https://github.com/AGLDWG/pid-proxy/blob/master/conf/0-linked.data.gov.au.conf) but they setup, as you've explained it here with root .htaccess and community/.htacces, with leading slashes removed, _should_ work.

Since http://qudt.org/community throws an error, are you sure you haven't got a pattern match for community already in your top-level.htaccess file which is then interacting with your community/ folder?

> in reading the documentation on githack, we should eventually be using the cdn URL out of respect for Pavel Puchkin, right?

Yes, that's right so once things are stable, we will do an update to that. Just want to see it up using the non-CDN link first, in case we have to make changes.

steveraysteveray commented:
OK, turns out there were some hard carriage returns in the file (I guess introduced by GitHub comments formatting). Here's the new version, which seems to work for the html, but not the embedded link to the Turtle. You can try it yourself.

```
RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule loop3d$ https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl [R=302,L]
RewriteRule loop3d.ttl$ https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.ttl [R=302,L]
RewriteRule loop3d$ https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.html [R=302,L]

RewriteCond %{QUERY_STRING} ^_mediatype=text/turtle$ [OR]
RewriteCond %{HTTP:Accept} text/turtle [NC]
RewriteRule loop3d/voc$ https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl [R=302,L]
RewriteRule loop3d/voc.ttl$ https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.ttl [R=302,L]
RewriteRule loop3d/voc$ https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/voc.html [R=302,L]
```

Suggested changes? 
How does Apache tell the difference from the first loop3d$ and the third loop3d$  line?

nicholascar commented:
@steveraysteveray I think all the redirects are working just fine!

* `curl http://qudt.org/community/loop3d` --> profile HTML
    * or just <http://qudt.org/community/loop3d> in browser
* `curl http://qudt.org/community/loop3d/voc` --> vocab HTML
    * or just <http://qudt.org/community/loop3d/voc> in browser
* `curl -L -H 'Accept: text/turtle' http://qudt.org/community/loop3d` --> profile RDF
* `curl -L -H 'Accept: text/turtle' http://qudt.org/community/loop3d/voc` --> vocab RDF
* `curl -L http://qudt.org/community/loop3d\?_mediatype=text/turtle`--> profile RDF
     *  or just <http://qudt.org/community/loop3d?_mediatype=text/turtle> in browser 
* `curl -L http://qudt.org/community/loop3d/voc\?_mediatype=text/turtle` --> profile RDF
      *  or just <http://qudt.org/community/loop3d/voc?_mediatype=text/turtle> in browser 

So it looks like everything's completed successfully! Good one figuring out the linebreaks... didn't occur to me!

> Suggested changes?

None!

> How does Apache tell the difference from the first loop3d$ and the third loop3d$ line?

`RewriteCond` lines apply to the next `RewriteRule` line so the two `RewriteCond` lines with the [OR] logical operator apply to the first loop3d$ line whereas the 3rd loop3d$ line is all by itself with no conditions. The upshot is that something matching loop3d$ needs to _also_ match one or other or both of the `RewriteCond` lines to redirect to the RDF result (.ttl) and if it does, the redirect matching concludes (due to [L] per `RewriteRule`). 

If no `RewriteCond` is matched, the matching will continue to loop3d.ttl$ and then the unqualified (no `RewriteCond` loop3d$ line.

Cheers, Nick

steveraysteveray commented:
I'd love to think we're done, but go to https://raw.githack.com/qudt/qudt-public-repo/master/community/loop3d/profile.html and click on the link to the Turtle file. 
That still doesn't work for me.

Same for the vocab.

nicholascar commented:
Oh, don't worry about that! _You_ are done (in terms of establishing the redirect etc.) but _I'm_ not! I'll have to update both HTML docs to ensure the RDF links work. Really, this is just me not being careful enough with the RDF links I requested and what I implemented!

I'll put in an HTML tidy-up PR next and I'll pop a README.md file into the `community/` folder too.

(I'm not too upset with myself as there is always content tidy-up stuff needed when persistent URIs get created!)

dr-shorthair commented:
For the record, I've never seen @nicholascar "too upset with himself" :-D

nicholascar commented:
Helpful @dr-shorthair, real helpful. Symptom of work "out in the open" these sorts of collegiate comments I suppose! :-)

nicholascar commented:
PR in to fix HTML links: PR #336

dr-shorthair commented:
My social interactions have been limited recently. Taking em where I can get em.
